File: backend1/Services/Implementations/UserService.cs
using backend.Exceptions;
using backend.Models;
using backend.Models.Entities;
using backend.Repositories;

namespace backend.Services.Implementations
{
    public class UserService : IUserService
    {
        private readonly IUserRepository _userRepository;

        public UserService(IUserRepository userRepository)
        {
            _userRepository = userRepository;
        }

        public async Task<APIResponse> GetUser(long userId)
        {
            var user = await _userRepository.GetByIdAsync(userId);

            if (user == null)
                throw new ResourceNotFoundException("User not found");

            return new APIResponse
            {
                Data = new UserDTO
                {
                    Id = user.Id,
                    Name = user.Name,
                    Email = user.Email,
                    Role = user.Role,
                    PhoneNumber = user.PhoneNumber,
                    Location = user.Location
                },
                Message = "User fetched successfully"
            };
        }

        public async Task<APIResponse> GetAllUsers()
        {
            var users = await _userRepository.GetAllAsync();
            var userDtos = users.Select(user => new UserDTO
            {
                Id = user.Id,
                Name = user.Name,
                Email = user.Email,
                Role = user.Role,
                PhoneNumber = user.PhoneNumber,
                Location = user.Location
            }).ToList();

            return new APIResponse
            {
                Data = userDtos,
                TotalElements = userDtos.Count,
                PageNumber = 1,
                PageSize = userDtos.Count,
                TotalPages = 1,
                Message = "Users fetched successfully"
            };
        }

        public async Task<APIResponse> UpdateUser(long id, UpdateUserRequest request)
        {
            var user = await _userRepository.GetByIdAsync(id);

            if (user == null)
                throw new ResourceNotFoundException("User not found");

            user.Name = request.Name;
            user.PhoneNumber = request.PhoneNumber;
            user.Location = request.Location;
            user.UpdatedAt = DateTime.UtcNow;

            await _userRepository.UpdateAsync(user);

            return new APIResponse
            {
                Data = new UserDTO
                {
                    Id = user.Id,
                    Name = user.Name,
                    Email = user.Email,
                    Role = user.Role,
                    PhoneNumber = user.PhoneNumber,
                    Location = user.Location
                },
                Message = "Profile updated successfully"
            };
        }

        public async Task<APIResponse> UpdatePassword(long id, UpdatePasswordRequest request)
        {
            var user = await _userRepository.GetByIdAsync(id);

            if (user == null)
                throw new ResourceNotFoundException("User not found");

            // Verify current password
            if (!BCrypt.Net.BCrypt.Verify(request.CurrentPassword, user.Password))
                throw new InvalidCredentialsException("Current password is incorrect");

            // Hash and update new password
            user.Password = BCrypt.Net.BCrypt.HashPassword(request.NewPassword);
            user.UpdatedAt = DateTime.UtcNow;

            await _userRepository.UpdateAsync(user);

            return new APIResponse
            {
                Message = "Password updated successfully"
            };
        }
    }
}

File: backend/src/main/java/com/wedspot/backend/services/implementation/UserService.java
package com.wedspot.backend.services.implementation;

import com.wedspot.backend.Model.APIResponse;
import com.wedspot.backend.Model.Entity.User;
import com.wedspot.backend.Model.UpdatePasswordRequest;
import com.wedspot.backend.Model.UpdateUserRequest;
import com.wedspot.backend.Model.UserDTO;
import com.wedspot.backend.exception.InvalidCredentialsException;
import com.wedspot.backend.exception.ResourceNotFoundException;
import com.wedspot.backend.mappers.UserMapper;
import com.wedspot.backend.repository.IUserRepository;
import com.wedspot.backend.services.IUserService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class UserService implements IUserService {

    private final IUserRepository userRepository;

    private final UserMapper userMapper;

    private final BCryptPasswordEncoder passwordEncoder;

    @Override
    public APIResponse getUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        UserDTO userDTO = userMapper.toDTO(user);
        APIResponse apiResponse = new APIResponse();
        apiResponse.setData(userDTO);
        apiResponse.setMessage("User fetched successfully");
        return apiResponse;
    }

    @Override
    public APIResponse getAllUsers() {
        List<User> users = userRepository.findAll();
        List<UserDTO> userDTOs = users.stream().map(userMapper::toDTO).collect(Collectors.toList());
        APIResponse apiResponse = new APIResponse();
        apiResponse.setData(userDTOs);
        apiResponse.setTotalElements(userDTOs.size());
        apiResponse.setPageNumber(0);
        apiResponse.setTotalPages(userDTOs.size());
        apiResponse.setMessage("Users fetched successfully");
        return apiResponse;
    }

    @Override
    public APIResponse UpdateUser(long id, UpdateUserRequest request) {
        User fetchedUser = userRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("User not found"));

        userMapper.updateEntityFromRequest(request, fetchedUser);

        User savedUser = userRepository.save(fetchedUser);

        APIResponse apiResponse = new APIResponse();
        apiResponse.setData(savedUser);
        apiResponse.setMessage("User updated successfully");
        return apiResponse;
    }

    @Override
    public APIResponse UpdatePassword(long id, UpdatePasswordRequest request) {
        User fetchedUser = userRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("User not found"));

        if(!fetchedUser.getPassword().equals(passwordEncoder.encode(request.getOldPassword()))) {
            throw new InvalidCredentialsException("Old password does not match");
        }

        fetchedUser.setPassword(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(fetchedUser);

        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("Password updated successfully");
        return apiResponse;
    }

    @Override
    public APIResponse deleteUser(long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        userRepository.delete(user);
        APIResponse apiResponse = new APIResponse();
        apiResponse.setMessage("User deleted successfully");
        return apiResponse;
    }
}
